rfc/egress-with-ucan.md

```diff
@@ -118,42 +141,67 @@ A UCAN delegation is used to allow token access to content. Typically, this is p
+        "aud": "did:web:ipfs.w3s.link",
+        // The Space
+        "sub": "did:key:zSpace",
+        "cmd": "/space/blob/get/0/1",
+        "args": {
+          // The `authToken` in the HTTP request, or...
+          "token": "the-token",
+          // null if none was given.
+          "token": null
+        },
+        // Random
+        "nonce": { "/": "abc123" },
+        "exp": null,
+        "prf": [
+           // The found delegations
+        ]
+      }
+      ```
+
+   4. Attempt to execute the invocation.
+      * If the invocation fails to authorize, respond with 401 Unauthorized.
+      * If authorization succeeds, execute as:
+        1. Perform the existing Gateway file retrieval, responding with the file data.
+        2. Asynchronously, inform the Accounting Service of the egress, to be billed to the Space.
+
+After deciding that a given token may retrieve a certain root CID from a certain Space, the Gateway MAY cache the decision by `(root-cid, token)` for some reasonable duration. The Gateway SHOULD NOT cache authorization *failures*, as the customer may be in the process of storing the delegation.
+
+## Hoverboard (Bitswap)
+
+Hoverboard does not handle tokens, but must respect the same authorization semantics as the Gateway does with a non-token request. Because there is no shareable service to perform that auth, Hoverboard will need to reimplement a simpler version of the same logic, only caring about the no-token execution path. This may be improved in the future, but some version of this is required to correctly implement private data storage.
+
+## Potential future work
+
+We may eventually create an object-store service. This service will sit between the Gateway and the Storage Nodes. It will speak UCAN, while the Storage Nodes will continue to speak only HTTP byte-range requests. This service will be the Executor of these `/space/blob/get/0/1` invocations, rather than the Gateway. This will allow other Principals to use the same mechanisms to retrieve Blobs, including Hoverboard as well as customers storing and retrieving Blobs directly. It would return a means to fetch the actual data in the receipt for the invocation, such as an HTTP URL.
 
 ## Open questions
 
-* What does the `/space/content/retrieve` receipt look like?
-* Can you enumerate the contents of a space? Is that in scope here?
-* What does a Gateway URLs with a token look like?
-* What can be cached? This seems relatively cacheable, but we should be explicit in the design to make sure we're on a suitable path. This process needs to be *fast*, at least once the cache is warm.
+* When auth fails, should we return 401 Unauthorized, or 404 Not Found to mask that the data exists?
+* Should `token` be `gatewayToken`? Or `authToken`? Is `token` too broad a term to be in the args of that invocation?
+* What if we find a Location Commitment, but it has no Space on it (since [it's optional](https://github.com/storacha/indexing-service/pull/18))?
```

Comment:
To be clear, once we begin using the indexing service, all new location commitments will have spaces on them.

It is optional only to account for data uploaded so far, which don't have a specific location commitment, and which we generate at query time for (i.e. to keep currently generated commitments valid) . If there is no space, we will assume limited public retrieval. (as if a null token delegation were stored)